### PR TITLE
feat: add new features available in flutter

### PR DESCRIPTION
**Pull request by yduartep**
1. New features Added:

- Offline Transport
- Data Collection Policy
- User Persistence
- Frame/Refresh Rate Monitoring
- Crash Reporting

2. Added configuration options for Frame Monitoring instrumentation (not available in Flutter SDK but default match Flutter). To enable frame monitoring, set `refreshRateVitals: true`.

- **targetFps**: Target frames per second for slow frame detection (default: 60)
- **frozenFrameThresholdMs**: Threshold in milliseconds for detecting frozen frames (default: 100ms)
- **refreshRatePollingInterval**: Interval in milliseconds for collecting and sending frame metrics
- **normalizedRefreshRate**: Normalized refresh rate for ProMotion displays (default: 60)

4. Include in DEMO app actions to replicate these new features if missing (default: 60)
5. Update to latest faro-core version `@grafana/faro-core@^2.2.3`.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,53 @@
+{
+  "version": "0.2.0",
+  "configurations": [
+    {
+      "name": "Debug Android (Hermes)",
```
Comment (robert-northmind) [inline .vscode/launch.json:5]: Just out of curiosity, what is this "Hermes" ?

File: demo/ios/FaroRNDemo/CrashTestModule.swift [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+/**
+ * Demo-only native module for testing crash reporting on iOS.
+ *
+ * This module is NOT part of the Faro SDK - it exists only in the demo app
+ * to allow triggering test crashes for demonstration purposes.
+ *
+ * WARNING: Do not include this in production apps!
+ */
+@objc(CrashTestModule)
```
Comment (robert-northmind) [inline demo/ios/FaroRNDemo/CrashTestModule.swift:11]: Is this how you build native components sort of in ReactNative..
So this CrashTestModule, can be triggered from the ReactNative code?

Should we think about using only pure swift if possible? easier to maintain than obj-c?
or the ReactNative logic is built around obj-c?

File: demo/src/faro/initialize.ts [hunk 3]
```diff
@@ -1,19 +1,22 @@
...
+import { initializeFaro } from '@grafana/faro-react-native';
+import type { ReactNativeConfig } from '@grafana/faro-react-native';
+
+/** In dev, use VERBOSE internal logger to diagnose collector connectivity issues */
+const FARO_DEBUG = __DEV__;
 
 /**
  * Demo versions to simulate different app releases
  */
 const DEMO_VERSIONS = ['1.0.0', '1.1.0', '1.2.0', '2.0.0', '2.1.0'];
```
Comment (robert-northmind) [inline demo/src/faro/initialize.ts:16]: Do these need to be hardcoded?
in the quick pizza flutter demo app, I think I just added a script which takes the current app version. and then choses to randomize 3 person of it.
It does this (if I remember correcT)

takes current version. and then either:
- keeps as is
- bumps patch with +1
- bumps patch with +2

e.g. if we have 1.2.0, then we randomize between
`1.2.0`, `1.2.1` and `1.2.2`

and if we even wanna bump the real version, we bump the minor.. so like `1.3.0`
then we know that minor is always a "real" new version

**Comment by robert-northmind** [comment]
Hey @yduartep ,
Great work on this. 🥳🤩

I had my cursor cloud agent have a look at this pr with some input from me and all looks great except for maybe these things:

Could fix:

makeRNConfig.ts and getRNInstrumentations.ts need test coverage. These are the composition roots that wire the entire SDK together — a bug here silently breaks everything downstream. At minimum, tests should cover:

Each config flag enables/disables the correct instrumentation
Default flags produce the expected instrumentation set
Transports are built correctly from url + enableTransports flags
Error is thrown when url is missing
Extra transports/instrumentations passed via config are appended correctly
Session tracking defaults merge as expected
sampling.ts needs tests. SamplingRate and SamplingFunction both have clamping logic (Math.max(0, Math.min(1, ...))) — simple unit tests to verify edge cases (negative values, values > 1, zero, boundary values) and context-based dynamic sampling.

Unused variable durationSoFar in both RefreshRateVitals.swift and FrameMonitor.kt — computed but never used in the slow frame event tracking branch. Remove or use it.
MemoryInfo.kt uses line!! (non-null assertion) — this violates the project's own coding guidelines in CLAUDE.md. Replace with safe access (val currentLine = line ?: continue or similar).
Thread safety on CPUInfo (both iOS and Android) — lastCpuTime / lastWallTime are static mutable state with no synchronization. If getCpuInfo() / getCpuUsage() is called concurrently from multiple threads, results will be incorrect. Consider synchronized / AtomicReference on Android, DispatchQueue or os_unfair_lock on iOS.
XHR instrumentation tests likely never execute. Two of four test cases bail early with if (typeof XMLHttpRequest === 'undefined') { return; }. In the Jest environment there's no real XMLHttpRequest, so these tests are effectively no-ops. Either mock XMLHttpRequest in the test setup or mark these as skipped with a TODO.

Nice to have:

UserDefaults.synchronize() is deprecated since iOS 12. It's called in FaroReactNative.persistSessionId() for crash resilience, but modern iOS auto-flushes frequently enough that it's unnecessary. Consider removing it to clear the deprecation warning.

File: .vscode/launch.json (hunk 1 above)
Comment (yduartep) [reply .vscode/launch.json:5]: Hermes engine is used to debug React Native apps. Use Chrome DevTools by navigating to chrome://inspect and selecting "Inspect" on your Hermes React Native target, or press 'j' in the Metro terminal. Hermes uses the Chrome Inspector protocol to enable breakpoint debugging, variable inspection, and console logging directly from your machine.

File: demo/ios/FaroRNDemo/CrashTestModule.swift (hunk 2 above)
Comment (yduartep) [reply demo/ios/FaroRNDemo/CrashTestModule.swift:11]: CrashTestModule is a native module to exposes methods, not a native UI component. Yes, JS can trigger it. In fact, the demo does that with NativeModules:

```
  const { CrashTestModule } = NativeModules;
  // ...
            if (CrashTestModule?.triggerTestCrash) {
              CrashTestModule.triggerTestCrash();
```

So the flow is: `JS → bridge → Objective-C surface → Swift implementation.`

**Why @objc and the .m file?**

React Native’s iOS bridge is built on the Objective-C runtime (RCTBridgeModule, macros like RCT_EXTERN_MODULE). Swift doesn’t see those APIs the same way, so you:

1. Mark the Swift class and methods with @objc so they’re visible to Obj-C.
2. Use a small Objective-C file to register the module and methods:

```
@interface RCT_EXTERN_MODULE(CrashTestModule, NSObject)

/**
 * Trigger a test crash (fatal error).
 * WARNING: This will immediately terminate the app!
 */
RCT_EXTERN_METHOD(triggerTestCrash)
```

**Maintainability**: keeping the real code in Swift and a minimal CrashTestModule.m is a common, reasonable approach. The Obj-C file is tiny, stable, and mostly declarations — not where most bugs live.

File: demo/src/faro/initialize.ts (hunk 3 above)
Comment (yduartep) [reply demo/src/faro/initialize.ts:16]: this is a random choice for the demo example app to send data to Grafana on different versions. Every time a new start of the app happens, it send a different version. That way we can build compare version panels in the plugin side for example.

**Comment by yduartep** [comment]
@robert-northmind Fixes applied:
- **sampling.test.ts** — sort-imports: use import { type SamplingContext, SamplingFunction, SamplingRate } from './sampling'.
- **setupTests.ts** — no-undef: removed queueMicrotask; mock XHR now completes synchronously in send() so events land in the same turn as send() (and avoids timing flakiness).
- **Root Prettier** — docs/mobile-rum/index.md was reformatted so quality:lint:root:prettier passes (it was failing before this run).
- **xhr/instrumentation.test.ts** — afterEach restores XMLHttpRequest.prototype.open/send to the baseline mock so repeated initializeFaro calls don’t stack patches and break pushEvent (stale instrumentation / nested originalSend).
The Jest message about a worker not exiting cleanly is still there; it’s a known pattern with open handles/timers somewhere in the suite and didn’t fail the run.

File: packages/react-native/src/config/types.ts [hunk 4]
```diff
@@ -1,91 +1,80 @@
...
+  frameMonitoringOptions?: FrameMonitoringOptions;
+  enableErrorReporting?: boolean;
+  enableCrashReporting?: boolean;
+  anrTracking?: boolean;
+  anrOptions?: ANRInstrumentationOptions;
+  enableConsoleCapture?: boolean;
+  consoleCaptureOptions?: Config['consoleInstrumentation'];
+  enableUserActions?: boolean;
+  userActionsOptions?: Config['userActionsInstrumentation'];
+  persistUser?: boolean;
```
Comment (robert-northmind) [inline packages/react-native/src/config/types.ts]: Not sure if this is planned as future work,
but we here seem to hint that we support auto persisting the user.
But I think the persistUser bool is not really used currently?

File: packages/react-native/README.md [hunk 5]
```diff
@@ -607,15 +672,23 @@ initializeFaro({
...
-    // Optional: Sampling rate (0-1) to sample sessions
+    enabled: true, // default: true
+    persistent: true, // default: false (volatile)
+    // Configurable timeouts (all in ms):
+    inactivityTimeout: 15 * 60 * 1000, // default: 15 min
+    sessionExpirationTime: 4 * 60 * 60 * 1000, // default: 4 h
+    maxSessionPersistenceTime: 15 * 60 * 1000, // default: 15 min
+
+    // Optional: Fixed sampling rate (0-1) to sample sessions
     samplingRate: 1.0, // 100% of sessions
```
Comment (robert-northmind) [inline packages/react-native/README.md]: I guess this old samplingRate does not apply anymore? since we now manage everything via the sampler? Should we just remove this one?

**Comment by yduartep** [comment]
Verification
persistUser only appeared on **ReactNativeConfig** in types.ts and in **docs/README examples**. It was never read in makeRNConfig, initializeFaro, or anywhere else, so it had no effect. User persistence in RN is handled separately via **initializeUserPersistence** / **createUserPersistence** (userPersistence/).

Changes
1. **packages/react-native/src/config/types.ts** — removed optional persistUser?: boolean from ReactNativeConfigOverrides.
2. **packages/react-native/README.md** — removed persistUser: true from the example initializeFaro block.
3. **docs/mobile-rum/configuration-comparison.md** — dropped persistUser from the RN snippet; updated the comparison table so RN is documented as using initializeUserPersistence() / UserPersistence, not a config flag.
4. **packages/react-native/README.md** — Removed samplingRate / sampler from the big example; noted optional sampling in comments. Session section now imports SamplingFunction / SamplingRate and documents them instead of the old keys.
5. **docs/mobile-rum/index.md** — Session sampling bullets, example, and comparison table now use SamplingRate / SamplingFunction; removed the old Flutter naming note that referenced samplingRate/sampler for RN.
docs/mobile-rum/configuration-comparison.md — Session sampling row updated for the sampling API.
6. **packages/react-native-tracing/README.md & CLAUDE.md** — Troubleshooting / performance snippets use new SamplingRate(...) and import from @grafana/faro-react-native.

File: packages/react-native/src/config/types.ts (hunk 4 above)
Comment (yduartep) [reply packages/react-native/src/config/types.ts]: removed

File: packages/react-native/README.md (hunk 5 above)
Comment (yduartep) [reply packages/react-native/README.md]: yes I removed
